drivers: clock_control: nrf: Use onoff service

nordic-krch opened the pull request:
Ported nRF clock control driver to use onoff service.
Clock_control API no longer tracks number of users but only enables or disables clock. It should not be used directly for clocks which have multiple user.

Clock control driver implements onoff service which can be used to track clock users. Clock control driver implements error detection mechanism which detects cases when direct API and onoff API is used for the same clock subsystem.

Nordic specific API (`nrf_clock_control.h`) provides function for getting onoff manager associated with given clock subsystem: `z_nrf_clock_control_get_onoff()`.

Since low frequency clock is typically used in enable once mode, nordic specific API provides a function for enabling LF clock without a need to maintain onoff client (`z_nrf_clock_control_lf_on()`). This is currently used by components which uses LF clock: system timer, bluetooth or counter driver.

PR contains port to use onoff service whenever HF clock is used.

PR is based on #23898 thus DNM.

Fixes #20712

nordic-krch committed:
drivers: clock_control: nrf: Add onoff service support

Updated nrf clock control driver to use onoff service for managing
multiple users.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
bluetooth: controller: Adapt to onoff clock control

Adapt link layer to use onoff service when controlling clock.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: usb: device: nrf: Adapt to control clock using onoff

Change to use onoff service when controlling HFXO.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
boards: arm: qemu_cortex_m0: Removed HFXO request from the timer

Removed HFXO request. It is a qemu target thus HF clock source
should not matter.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: led_strip: ws2812_gpio: Adapt to clock using onoff

Adapted to use onoff service for clock control

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: clock_control: Add suite for onoff with clock

Added suite to test clock control with onoff.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: clock_control: nrf: Add clock status shell command

Added shell command to clock driver to get clock status.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
manifest: Refer to updated nrf_hw_models

nrf_hw_models require extension to support nrf_clock_is_running call.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -118,100 +111,69 @@ static const struct nrf_clock_control_sub_config *get_sub_config(
-
-	return err;
+	return (GET_STATUS(*flags) != GET_STATUS(prev_flags)) ?
+			-EFAULT : -EPERM;
```
This seems to be undocumented magic.  Why these two values?  What do they mean?  They propagate to be return values to clock_control_stop(), and require modifications to the generic clock_control API to invoke Nordic-specific recovery, so they're essentially public API, but neither is documented.

nordic-krch replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/24334#discussion_r452757361

pabigot reviewed (commented):
I don't really understand all that's going on here.  I guess the overall summary is:
* The existing clock_control API doesn't work for Nordic;
* A new mostly-independent Nordic-specific clock control based on onoff is added;
* There's some effort to make some standard clock_control API work using the new Nordic-specific API, though it appears to require callers to interpret undocumented return codes and do special cleanup in some cases;
* Some examples/drivers are switched to use the Nordic-specific API, which is OK because the examples only work on Nordic anyway.

I was hoping to see the current clock_control API implemented in terms of onoff without affecting existing users.  If the existing API doesn't meet functional requirements (as when `clock_control_off()` can't be made non-blocking, though IIRC for Nordic clocks it is) then the API should be extended.

I don't think that most code that currently uses clock_control should need to know that a particular driver implements the API using onoff underneath.  But without digging deep into the implementation I can't say definitively that such an approach could work, only that this PR doesn't do it that way.

I think my response to this PR will be abstain.

nordic-krch commented:
@pabigot
 
> The existing clock_control API doesn't work for Nordic;

The existing clock_control API was used differently by nordic and others. That is because majority of clocks have single user and does not need reference counting, etc. Nordic was using it differently, with clock control. We also added `struct clock_control_async_data` to asynchronous API to allow queueing requests.

With onoff ready, my idea is to simplify clock control API to be single user only. Clocks that may have multiple users must not use clock control API directly but rather access it using associated onoff service which is then the single user of the driver. I prefer that over expanding clock control API.

Also note that for nrf53 there are clocks which has only single user (e.g. dedicated 48Mhz USB clock) so even not all clock subsystems in nordic will have onoff service.

>There's some effort to make some standard clock_control API work using the new Nordic-specific API

API can be extended with that error codes.

The reality is that clock control API is used by vendor specific code since clock subsystem are vendor specific so standardize API does not provide portability but rather standardize pattern. Also decision whether use API directly or onoff requires specific knowledge.

pabigot commented:
@nordic-krch please rebase and remove DNM

carlescufi commented on include/drivers/clock_control/nrf_clock_control.h:
```diff
@@ -94,4 +102,22 @@ int z_nrf_clock_calibration_count(void);
+ *
+ * @return Service handler or NULL.
+ */
+struct onoff_manager *z_nrf_clock_control_get_onoff(clock_control_subsys_t sys);
```
@tbursztyka @mnkp @jdascenzio @jfischer-phytec-iot @galak @mbolivar-nordic feedback welcome on this new pattern: obtaining an instance to the on/off service through the existing driver API.

tbursztyka replied:
I did not see any changes on the public API side. Instead, it exposes private features to this driver. I don't have much opinion on this then. If that's solving an important nordic use case but has no grounds for being generalized, the there is no other way. Fine by me.

mbolivar-nordic replied:
I agree. At dev-review there was no feedback that others had similar requirements, so it seems appropriate to do it this way, at least to get started.

pabigot replied:
Yes.  The original design concept did not expect these structures to be exposed, but in this case it's the most sensible approach.  The expected alternative (provide custom request/release API specific to the subsystem) can be demonstrated where there's value added by the wrappers.

nordic-krch replied:
#22409 had the proposal to make it more generic.

nordic-krch commented:
@aescolar can you help me with btsim test that i have failing here. 

https://buildkite.com/zephyr/zephyr/builds/1010#863e82cf-ec84-4a7c-b73b-edb1120ccfd0
```

tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_encrypted_split.sh PASSED
--
  | tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_encrypted_split_privacy.sh PASSED
  | tests/bluetooth/bsim_bt/bsim_test_advx/tests_scripts/basic_advx.sh PASSED
  | tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_split.sh PASSED
  | tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/gap.sh FAILED
  | d_01: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.3.0-500-g542e26ed4e61  ***
  | d_02: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.3.0-500-g542e26ed4e61  ***
  | d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_raw: Bluetooth enabled in RAW mode
  | d_02: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_raw: Bluetooth enabled in RAW mode
  | Traceback (most recent call last):
  | File "/var/lib/buildkite-agent/zephyr-module-cache/tools/edtt/src/tests/gap_verification.py", line 1934, in run_a_test
  | success = success and test_f(transport, 0, 1, trace);
  | File "/var/lib/buildkite-agent/zephyr-module-cache/tools/edtt/src/tests/gap_verification.py", line 1406, in gap_adv_bv_01_c
  | success = advertiser.enable();
  | File "/var/lib/buildkite-agent/zephyr-module-cache/tools/edtt/src/components/advertiser.py", line 160, in enable
  | success = success and self.__advertise_enable(Advertise.ENABLE);
  | File "/var/lib/buildkite-agent/zephyr-module-cache/tools/edtt/src/components/advertiser.py", line 146, in __advertise_enable
  | self.status = le_set_advertising_enable(self.transport, self.idx, enable, 100);
  | File "/var/lib/buildkite-agent/zephyr-module-cache/tools/edtt/src/components/basic_commands.py", line 1073, in le_set_advertising_enable
  | raise Exception("LE Set Advertising Enable command failed: Response too short (Expected %i bytes got %i bytes)" % (5, len(packet)));
  | Exception: LE Set Advertising Enable command failed: Response too short (Expected 5 bytes got 0 bytes)
  | Traceback (most recent call last):
  | File "/var/lib/buildkite-agent/zephyr-module-cache/tools/edtt/src/components/preambles.py", line 87, in pream

```
Pull request is changing clock control, starting HF and have no idea if that can be the reason. When tried locally it also fails also on master.

nordic-krch commented:
@anangl @pabigot can you review? CI is starting to look ok (except for bsim tests which seems to fail the same way on master, at least on my setup).

nordic-krch commented:
@emob-nordic can you take a look at USB changes?

nordic-krch commented:
@joerchan can you take a look at bluetooth changes?

aescolar commented on west.yml:
```diff
@@ -123,7 +123,7 @@ manifest:
     - name: nrf_hw_models
       path: modules/bsim_hw_models/nrf_hw_models
-      revision: fec69703cb1ca06fcdab6d5fde01274f0fc5c759
+      revision: 3e9ee62072c8b45a2596ee9e08f12e6c978ae44a
```
Please point to the latest: 9e594dace1af29252903938064b8ecb1b8b77678 (no reason to point to older versions)

aescolar commented:
@nordic-krch 

> When tried locally it also fails also on master.

master seems fine. Are your BT changes working on real HW?

nordic-krch commented:
> master seems fine. Are your BT changes working on real HW?

@aescolar  I got to the point that in fact master is passing but this branch is not. I tried to add some logging to see if hf clock is controlled ok and it seems ok. I'm not sure what can be the issue then. I see that for example `hci.sh` is failing on not receiving a message
```
raise Exception("Flush Events command failed: Response too short (Expected %i bytes got %i bytes)" % (4, len(packet)));
Exception: Flush Events command failed: Response too short (Expected 4 bytes got 0 bytes)
```
Have you experienced such errors? Could that somehow be clock related? Logs show no asserts or errors.

aescolar commented:
@joerchan @cvinayak Do you have some idea about what could have broken?

emob-nordic commented:
> @emob-nordic can you take a look at USB changes?

Sorry for the delay. I will have a look.

joerchan commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -486,7 +486,7 @@ void lll_isr_cleanup(void *param)
 
 	err = lll_hfclock_off();
-	LL_ASSERT(!err || err == -EBUSY);
+	LL_ASSERT(err >= 0);
```
Why accept positive error values?

nordic-krch replied:
it is using onoff service underneath which returns non-negative for success: https://github.com/zephyrproject-rtos/zephyr/blob/979e124e21fb05cfedf8ed12078983d40e2f0b6c/include/sys/onoff.h#L338

joerchan commented:
> @joerchan can you take a look at bluetooth changes?

@nordic-krch I'm not the right person to look at the Bluetooth controller changes. I'm surprised the code-owners file did not pick them up, but I have added @cvinayak and @kruithofa instead.

anangl commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -25,40 +25,50 @@ LOG_MODULE_REGISTER(temp_nrf5, CONFIG_SENSOR_LOG_LEVEL);
-	r = clock_control_async_on(data->clk_dev, CLOCK_CONTROL_NRF_SUBSYS_HF,
-					&clk_data);
-	__ASSERT_NO_MSG(!r);
+	onoff_client_init_callback(&cli, hfclk_on_callback, &do_start);
```
Changes to `temp_nrf5.c` should be extracted from this commit and merged with those in commit 3b13686324e1a6a06041294453517bca8a838985 (drivers: sensor: nrf5: Adapt temp driver to use onoff for clock), because currently they use no longer existing onoff API.

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -186,18 +187,43 @@ void rtc_nrf_isr(void *arg)
 }
 
-int z_clock_driver_init(struct device *device)
+static void clock_on_notify(struct onoff_manager *mgr,
```
Most of the changes introduced here are removed in commit f2d1f8297f15bc4cffcb78bf7fd92700e339d7df (drivers: timer: nrf: Modify LF clock request). Maybe it would be better to change this file in one step?

nordic-krch replied:
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -226,68 +188,157 @@ static inline void anomaly_132_workaround(void)
+	nrf_clock_lfclk_t type;
+
+	while ((nrf_clock_is_running(NRF_CLOCK, d, (void *)&type) == false)
+		&& (type != t)) {
```
no, i want to block until requested type returns true. For example `nrf_clock_is_running` for HF clock will always return true but with RC source.

anangl replied:
> i want to block until requested type returns true

That's exactly how I've read your intention. You need then:
```
while (!(nrf_clock_is_running(NRF_CLOCK, d, (void *)&type) && (type == t)))
```
what is equivalent to what I mentioned in my previous comment.

nordic-krch replied:
Stupid me. Fixed.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -226,68 +188,157 @@ static inline void anomaly_132_workaround(void)
 
-	if (ref == 1) {
-		DBG(dev, subsys, "Triggering start task");
+	return k_sem_take(&sem, K_MSEC(500));
```
`clock_control_on` in 'clock_control.h' states
```
This function may sleep, and thus can only be called from thread context.
```
For me it means that it can only be called from app which supports multithreading.

nordic-krch replied:
We could extend description to state that in non sleepable context it just spins and then extend implementation.

anangl replied:
I think that "from thread context" in that description is supposed to mean "not from ISR". And someone may want to call this function from `main()` in an app without `MULTITHREADING` enabled. An attempt to call `k_sem_take` with a parameter other than `K_NO_WAIT` will end up in UsageFault then (I thought it would just not wait but apparently I was wrong). I think this case should be handled somehow, perhaps with a loop calling `k_cpu_idle()` while waiting for this semaphore.

nordic-krch replied:
just learn that interrupts does not work if multhreading is off (#26796). I will return error in that case.

nordic-krch committed:
drivers: sensor: nrf5: Adapt temp driver to use onoff for clock

Adapted driver to use onoff service for clock control.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: timer: nrf: Adapt to clock_control changes

Adapted driver to clock_control changes (usage of onoff manager).
Since timer is permanenty requesting the clock, it is using API
dedicated for that: z_nrf_clock_control_lf_on().

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: nrfx_rtc: Adapt to clock control changes

Adapted nrfx_rtc counter to request clock through onoff API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
bsim tests finally fixed. Build is green. Can you finish the review?

carlescufi commented:
@cvinayak please review the Bluetooth commit: https://github.com/zephyrproject-rtos/zephyr/pull/24334/commits/68f6af0a4973fea88fef7112fb63631286b9400d

@emob-nordic Please review the USB changes.

anangl commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -523,62 +526,33 @@ void usb_dc_nrfx_power_event_callback(nrf_power_event_t event)
+static int hfxo_start(struct nrf_usbd_ctx *ctx)
+{
+	if (atomic_cas(&ctx->clk_single_req, 0, 1)) {
+		sys_notify_init_callback(&ctx->hfxo_cli.notify, hfxo_rdy_cb);
```
Wouldn't `sys_notify_init_spinwait()` be sufficient here, is this empty callback needed for something?

anangl commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -231,7 +256,8 @@ void z_nrf_clock_calibration_init(struct device *dev)
 	nrf_clock_event_clear(NRF_CLOCK, NRF_CLOCK_EVENT_DONE);
 	nrf_clock_int_enable(NRF_CLOCK, NRF_CLOCK_INT_DONE_MASK);
 
-	clk_dev = dev;
+	hf_mgr = z_nrf_clock_control_get_onoff(CLOCK_CONTROL_NRF_SUBSYS_HF);
+	lf_mgr = z_nrf_clock_control_get_onoff(CLOCK_CONTROL_NRF_SUBSYS_LF);
```
The `dev` parameter is no longer used. Instead of it, you could pass to `z_nrf_clock_calibration_init()` the pointer to the onoff_manager array, and actually store only this one pointer.

pabigot commented on drivers/led_strip/ws2812_gpio.c:
```diff
@@ -140,7 +146,7 @@ static int send_buf(struct device *dev, uint8_t *buf, size_t len)
 	irq_unlock(key);
 
-	rc = clock_control_off(clk, CLOCK_CONTROL_NRF_SUBSYS_HF);
+	rc = onoff_release(mgr);
```
This propagates to a behavior change to `led_strip_update_rgb()`: `clock_control_off` returns zero on success; `onoff_release()` may return a positive value on success.

nordic-krch replied:
fixed

pabigot commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -523,62 +526,33 @@ void usb_dc_nrfx_power_event_callback(nrf_power_event_t event)
+static int hfxo_start(struct nrf_usbd_ctx *ctx)
+{
+	if (atomic_cas(&ctx->clk_single_req, 0, 1)) {
+		sys_notify_init_callback(&ctx->hfxo_cli.notify, hfxo_rdy_cb);
```
Since the callback does nothing why not use a spinwait notify here?

nordic-krch replied:
removed.

pabigot commented on include/drivers/clock_control/nrf_clock_control.h:
```diff
@@ -28,6 +30,13 @@ enum clock_control_nrf_type {
+/** @brief LF clock start modes. */
+enum nrf_lfclk_start_mode {
+	NRF_LFCLK_START_MODE_NOWAIT,
+	NRF_LFCLK_START_MODE_SPINWAIT_STABLE,
```
If RUNNING occurs before STABLE it would be nice to have that earlier in the enumeration, so magnitude corresponds to functionality.

nordic-krch replied:
done

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -226,68 +188,151 @@ static inline void anomaly_132_workaround(void)
-	if ((data != NULL)
-	    && is_in_list(&clk_data->list, &data->node)) {
-		return -EBUSY;
+	err = set_state(&subdata->flags, CLOCK_CONTROL_STATUS_OFF,
+			ctx | CLOCK_CONTROL_STATUS_STARTING);
+	if (err < 0) {
+		return err;
```
AFAICT this will return `-EFAULT` if an already started (or starting) clock is started again by a call to `clock_control_on()` or `clock_control_async_on()` (or maybe even `-EPERM` if `subdata->flags` for some reason got changed between the call to `atomic_cas()` and later comparison of the flags, what seems to be not impossible). Previous implementation returned 0 in such case. Is that an intentional change in behavior?

nordic-krch replied:
i've rewrote setting state. `-EBUSY` is returned if on is requested (blocking or async) and clock is not in off state. We need clearance in the API. Since we are removing reference counting from clock_control API then second on request should always error (`-EBUSY`).

The problem is the order. I would prefer that this PR goes in which will violate API in `clock_control_async_on` by returning `-EBUSY` when clock is already running (API states that it should return 0) but:
- it is used only by nordic.
- nordic will no longer use it directly after switch to onoff service.

anangl replied:
> We need clearance in the API. Since we are removing reference counting from clock_control API then second on request should always error (`-EBUSY`).

I'm not convinced we should do it. What about other clock_control drivers? They don't seem to check if the clock was already requested. Should they all be updated then?
But why actually the second request should be considered an error?

nordic-krch replied:
> But why actually the second request should be considered an error?

Since `calling clock_control_off()` shuts down the clock (without reference counting) then imo it should be only one originator of clock being on. Wouldn't it be error prone? If there are multiple sources which turns clock on and off then it calls for onoff and it might go unnoticed if `clock_control_on` returns success when enabling already enabled clock. imo, on and off should return  `-EALREADY` (better than `-EBUSY`) when clock is in unexpected state. 

> What about other clock_control drivers? 

Note that api clock control is still poor defined with no description of errors in `clock_control_on` and `clock_control_off`. Another thing is that typically (always?) clock control is used by vendor dependent layer so it is not to give hardware abstraction but rather common look and feel (and test suite).

anangl replied:
Agreed. This will indeed help identify cases where switching to using the on-off service is needed. And yes, `-EALREADY` seems more suitable for both `clock_control_on()` and `clock_control_off()`. But I don't think it will help if we mention that in the API description. Let's stay with this opaque "0 on success, negative errno on failure". This allows more flexible approach to be used by different vendors, what is actually good for such vendor specific thing like clock handling.

pabigot replied:
If you want opaque and flexible use non-negative for success so vendors can convey additional information (like whether the request was unnecessary) in the non-error return value.

pabigot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -82,43 +96,26 @@ int lll_hfclock_on(void)
-	k_sem_give(user_data);
+	err = onoff_cancel_or_release(hf_mgr, &hf_state.cli);
+	/* Clean notify state. It is currently not clean during cancellation. */
+	hf_state.cli.notify.flags = SYS_NOTIFY_METHOD_COMPLETED;
```
Why do you need to "clean" the notify state?

nordic-krch replied:
removed.

pabigot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -19,61 +19,75 @@
+	k_sem_give(&clk_state->sem);
+}
+
+int clk_request(struct onoff_manager *mgr, struct lll_clock_state *state)
```
This should be `static` as it's currently a global symbol.

nordic-krch replied:
done

pabigot reviewed (changes requested):
Github seems to be losing comments made on different files, so I'm recording this one while I look to see where it put the others.

pabigot commented:
@nordic-krch Please respond to https://github.com/zephyrproject-rtos/zephyr/pull/24334/files#r409713202 which relates to https://github.com/zephyrproject-rtos/zephyr/pull/24334#discussion_r451602336 from @anangl.

That code is also reading an atomic data structure without using the atomic API.  It seems inherently racy.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -82,43 +96,26 @@ int lll_hfclock_on(void)
-	DEBUG_RADIO_XTAL(1);
-
-	return err;
+	return on_wait(&hf_state, HFCLOCK_TIMEOUT_MS);
```
`lll_hfclock_on_wait` functionality is changed. This function needs to start hfclock in blocking mode. In the change it is only doing `k_sem_take` followed by `k_sem_give` in the `on_wait` function.

nordic-krch replied:
right, misread it, i thought it's blocking waiting for already request clock. fixed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -19,61 +19,75 @@
-	if (done) {
-		return 0;
+	if (err == 0) {
+		k_sem_give(&state->sem);
```
Why give after a take?

nordic-krch replied:
rewrote it, initially, i didn't want to add another request to lf clock but since it is permanently requested it is much easier to add new blocking request.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -19,61 +19,75 @@
+	struct lll_clock_state *clk_state =
+		CONTAINER_OF(cli, struct lll_clock_state, cli);
+
+	k_sem_give(&clk_state->sem);
```
semaphore is not needed for non-blocking calls.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -19,61 +19,75 @@
+}
 
-	k_sem_init(&sem_clock_wait, 0, 1);
+int lll_clock_init(void)
```
`lll_clock_init` is non-blocking LFCLOCK request interface. Semaphores should not be given.

nordic-krch replied:
rewrote. Please check.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -19,61 +19,75 @@
-	err = k_sem_take(&sem_clock_wait, K_MSEC(LFCLOCK_TIMEOUT_MS));
 
-	return err;
+int lll_clock_wait(void)
```
`lll_clock_wait` is blocking LFCLOCK request. Here, now, only semaphores are taken and given?

nordic-krch replied:
fixed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -82,43 +96,26 @@ int lll_hfclock_on(void)
-	} else if (err == -EBUSY) {
-		DEBUG_RADIO_XTAL(1);
+	if (hf_refcnt < 1) {
+		return -EAGAIN;
```
`-EALREADY` HFCLOCK is released.

nordic-krch replied:
changed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -82,43 +96,26 @@ int lll_hfclock_on(void)
 
-	return err;
-}
+	if (atomic_dec(&hf_refcnt) > 1) {
```
```suggestion
	if (atomic_dec(&hf_refcnt) > 0) {
```

nordic-krch replied:
`atomic_dec` returns previous value. So 1 means that refcnt reached 0.

emob-nordic commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -251,6 +251,9 @@ K_MEM_POOL_DEFINE(ep_buf_pool, EP_BUF_POOL_BLOCK_MIN_SZ,
  */
 struct nrf_usbd_ctx {
 	usb_dc_status_callback status_cb;
+	struct onoff_client hfxo_cli;
+	struct onoff_manager *hfxo_mgr;
+	atomic_t clk_single_req;
```
Please provide some description for those as it is done for other fileds in this struct. Also Could You please rename `clk_single_req` back to `clock_requested`? I assume it has same purpose but older name says more to me.

pabigot commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -524,3 +536,47 @@ void nrf5_power_usb_power_int_enable(bool enable)
+	irq_unlock(key);
+
+	shell_print(shell, "HF clock:");
+	shell_print(shell, "\t- %srunning (users: %d)",
```
Another comment that hasn't been addressed.

nordic-krch replied:
done

emob-nordic commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -242,6 +242,9 @@ K_MEM_POOL_DEFINE(ep_buf_pool, EP_BUF_POOL_BLOCK_MIN_SZ,
  * @param status_cb	Status callback for USB DC notifications
+ * @param hfxo_cli	Onoff client used to control HFXO
+ * @param hfxo_mgr	Pointer to onoff manager associated with HFXO.
+ * @param clk_reuqested	Flag used to protect against double stop.
```
Typo.
s/clk_reuqested/clk_requested

nordic-krch replied:
fixed

nordic-krch commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -5066,10 +5075,7 @@ static inline void isr_radio_state_close(void)
-	if (!err) {
-		DEBUG_RADIO_XTAL(0);
-	}
+	hf_xtal_stop();
```
this file is removed, right?

nordic-krch commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -61,16 +61,49 @@ static void setup_instance(const char *dev_name, clock_control_subsys_t subsys)
 				clock_control_subsys_t subsys)
 {
-	struct device *dev = device_get_binding(dev_name);
+#ifdef DT_INST_0_NORDIC_NRF_CLOCK_LABEL
```
done

carlescufi commented:
@nordic-krch can you please take another look at the comments?

nordic-krch commented on tests/drivers/clock_control/onoff/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+  drivers.clock.clock_control_onoff:
+    tags: drivers
+    platform_whitelist: nrf51_pca10028 nrf52_pca10040 nrf52840_pca10056
+                        nrf9160_pca10090
```
done

carlescufi commented:
@anangl @cvinayak please re-review

anangl reviewed (approved):
I'm still not entirely convinced about the values that `clock_control_on()` and `clock_control_off()` return when the clock is not in the expected state. Currently, these functions return `-EBUSY` and `0`, respectively. It seems more suitable to use `-EALREADY` in both cases, as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/24334#discussion_r453592208. But it's a minor thing. No reason to further block this PR from my side.

koffes commented on tests/drivers/clock_control/onoff/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+tests:
+  drivers.clock.clock_control_onoff:
+    tags: drivers
+    platform_whitelist: nrf51dk_nrf51422 nrf52dk_nrf52832 nrf52840dk_nrf52840
```
Jumping in from the sideline: What about 5340?

cvinayak commented:
Profile CPU usage on nRF51:

**Upstream master 6e0c4c79bc samples/bluetooth/peripheral advertising CPU usage:**
![image](https://user-images.githubusercontent.com/6350656/87536826-020ea980-c69a-11ea-96c1-c574f2f102bf.png)

**This PR CPU usage:**
![image](https://user-images.githubusercontent.com/6350656/87536790-f622e780-c699-11ea-909b-f4ecf5b9d43f.png)

**Comments:**
There is a significantly noticeable degradation in performance/efficiency due to increased CPU use due to onoff service compared to earlier clock_control implementation.
BLE event preparation that turns on HFClock asynchronously sees an increase of **~26us** (0.59625ms - 0.5705625ms), and on event close the increase is **~31us** (0.1950625ms - 0.163875).
To compare, a typical Radio Rx ISR (header complete timeout) in connectable advertising state takes ~62us.

cvinayak reviewed (changes requested):
Code changes in BLE controller look good, except for the significant degradation in performance.

nordic-krch commented:
> except for the significant degradation in performance.

@cvinayak thanks for checking this. I will take a look what is the onoff overhead and what can be done about it.

nordic-krch committed:
tests: drivers: clock_control: nrf: Add suite for bluetooth API

Nordic platforms exposes API for controlling HF clock along
onoff API. This API is dedicated to be used by the bluetooth
controller and must not be used elsewhere.

Test suite is validating stability of HF clock control using
onoff API and specialized API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@cvinayak can you take one more look? I've added `void z_nrf_clock_bt_ctrl_hf_request(void)` and `void z_nrf_clock_bt_ctrl_hf_release(void)` to be used by the controller.

@anangl can you take another look at the driver implementation as i had to add some hacks&flags to support out of band hf clock requesting. I've also added a test suite where i tried to mix ble and onoff API and stress test it.

Overall, that didn't make clock control driver prettier so I hope that this backdoor was opened for only one guest.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -16,6 +17,20 @@ LOG_MODULE_REGISTER(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+#define CTX_API			BIT(7)
+#define CTX_MASK (CTX_ONOFF | CTX_API)
+
+/* Used only by HF clock */
+#define HF_BT_ACTIVE		BIT(8)
+#define HF_GENERIC_ACTIVE	BIT(9)
+#define HF_USER_MASK		(HF_BT_ACTIVE | HF_GENERIC_ACTIVE)
```
Since this is only for HFCLK, why not use a dedicated variable for it? Something like:
```
static atomic_t hfclk_users;
```
You would then avoid the need of this weird masking of `flags` with `HF_USER_MASK` in `set_off_state()`:
https://github.com/zephyrproject-rtos/zephyr/blob/a7fbf5e80f3f702024c261dd63899f368f39028d/drivers/clock_control/nrf_power_clock.c#L158
and in `set_starting_state()`:
https://github.com/zephyrproject-rtos/zephyr/blob/a7fbf5e80f3f702024c261dd63899f368f39028d/drivers/clock_control/nrf_power_clock.c#L173-L175
Also accesses to such variable should be faster than to `flags` for which you need to get the pointer via `get_hf_flags()`.

anangl commented on include/drivers/clock_control/nrf_clock_control.h:
```diff
@@ -94,4 +103,42 @@ int z_nrf_clock_calibration_count(void);
+ * ensure that every z_nrf_clock_bt_ctrl_hf_request matches
+ * z_nrf_clock_bt_ctrl_hf_release call.
+ */
+void z_nrf_clock_bt_ctrl_hf_request(void);
```
You should rather use `*_bt_ctlr_*`, for consistency with other related stuff.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -226,68 +220,277 @@ static inline void anomaly_132_workaround(void)
+
+	*flags |= HF_GENERIC_ACTIVE;
+	if (*flags & HF_BT_ACTIVE) {
+		nrf_clock_is_running(NRF_CLOCK, NRF_CLOCK_DOMAIN_HFCLK, &type);
```
The value returned by `nrf_clock_is_running()` is ignored here, so `(void)` could help make it clear (and perhaps also avoid warnings from some linters).

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -6,16 +6,32 @@
+/* Used only by HF clock */
+#define HF_USER_BT		BIT(0)
+#define HF_USER_GENERIC		BIT(1)
+#define HF_USER_MASK		(HF_USER_BT | HF_USER_GENERIC)
```
No longer needed.

nordic-krch replied:
removed.

nordic-krch commented:
@cvinayak can you check bluetooth part?

cvinayak commented:
PR:
![image](https://user-images.githubusercontent.com/6350656/87934660-c3a92e00-ca8f-11ea-9810-a5a4a6bcc7ae.png)

Master 98bcc51b09:
![image](https://user-images.githubusercontent.com/6350656/87934743-ec312800-ca8f-11ea-81fd-0c6d114ed603.png)

The performance is now comparable.

cvinayak reviewed (approved):
Approved for BLE controller changes by analyzing performance changes.

ioannisg commented on boards/arm/qemu_cortex_m0/nrf_timer_timer.c:
```diff
@@ -162,18 +162,8 @@ void timer0_nrf_isr(void *arg)
-
-	clock_control_on(clock, CLOCK_CONTROL_NRF_SUBSYS_HF);
-
-	/* FIXME switch to 1 MHz once this is fixed in QEMU */
```
You should leave the FIXME line, this is about an issue in QEMU

nordic-krch replied:
comment is back
